pulsar/impl_consumer.go

```diff
@@ -0,0 +1,256 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package pulsar
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/apache/pulsar-client-go/pkg/pb"
+	"github.com/apache/pulsar-client-go/util"
+	"github.com/golang/protobuf/proto"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type consumer struct {
+	topicName       string
+	consumers       []Consumer
+	log             *log.Entry
+	queue           chan ConsumerMessage
+	unackMsgTracker *UnackedMessageTracker
+}
+
+func newConsumer(client *client, options *ConsumerOptions) (*consumer, error) {
+	if options == nil {
+		return nil, newError(ResultInvalidConfiguration, "consumer configuration undefined")
+	}
+
+	if options.Topic == "" && options.Topics == nil && options.TopicsPattern == "" {
+		return nil, newError(TopicNotFound, "topic is required")
+	}
+
+	if options.SubscriptionName == "" {
+		return nil, newError(SubscriptionNotFound, "subscription name is required for consumer")
+	}
+
+	if options.ReceiverQueueSize == 0 {
+		options.ReceiverQueueSize = 1000
+	}
+
+	if options.TopicsPattern != "" {
+		if options.Topics != nil {
+			return nil, newError(ResultInvalidConfiguration, "Topic names list must be null when use topicsPattern")
+		}
+		// TODO: impl logic
```

Comment:
create an issue and add the link here. then logFatal since it is not implemented yet.

Reply:
The function will impl in https://github.com/apache/pulsar-client-go/issues/32